lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+  KEY_ATTRIBUTES = 'attributes'.freeze
+  KEY_CHECKS = 'checks'.freeze
+  KEY_EQUALITY = 'equality'.freeze
+  KEY_EQUALITY_INCLUDE_TYPE = 'equality_include_type'.freeze
+  KEY_FINAL = 'final'.freeze
+  KEY_FUNCTIONS = 'functions'.freeze
+  KEY_KIND = 'kind'.freeze
+  KEY_OVERRIDE = 'override'.freeze
+  KEY_PARENT = 'parent'.freeze
+  KEY_TYPE = 'type'.freeze
+  KEY_VALUE = 'value'.freeze
+
+  ATTRIBUTE_KIND_CONSTANT = 'constant'.freeze
+  ATTRIBUTE_KIND_DERIVED = 'derived'.freeze
+  ATTRIBUTE_KIND_GIVEN_OR_DERIVED = 'given_or_derived'.freeze
+  TYPE_ATTRIBUTE_KIND = TypeFactory.enum(ATTRIBUTE_KIND_CONSTANT, ATTRIBUTE_KIND_DERIVED, ATTRIBUTE_KIND_GIVEN_OR_DERIVED)
+
+  TYPE_ANNOTATION_KEY_TYPE = PType::DEFAULT # TBD
+  TYPE_ANNOTATION_VALUE_TYPE = PStructType::DEFAULT #TBD
+  TYPE_ANNOTATIONS = PHashType.new(TYPE_ANNOTATION_KEY_TYPE, TYPE_ANNOTATION_VALUE_TYPE)
+
+  TYPE_MEMBER_NAME = PPatternType.new([PRegexpType.new(Patterns::PARAM_NAME)])
+
+  TYPE_ATTRIBUTE = TypeFactory.struct({
+    KEY_TYPE => PType::DEFAULT,
+    KEY_ANNOTATIONS => TypeFactory.optional(TYPE_ANNOTATIONS),
+    KEY_FINAL => TypeFactory.optional(PBooleanType::DEFAULT),
+    KEY_OVERRIDE => TypeFactory.optional(PBooleanType::DEFAULT),
+    KEY_KIND => TypeFactory.optional(TYPE_ATTRIBUTE_KIND),
+    KEY_VALUE => PAnyType::DEFAULT
+  })
+  TYPE_ATTRIBUTES = TypeFactory.hash_kv(TYPE_MEMBER_NAME, TypeFactory.not_undef)
+
+  TYPE_FUNCTION_TYPE = PType.new(PCallableType::DEFAULT)
+
+  TYPE_FUNCTION = TypeFactory.struct({
+    KEY_TYPE => TYPE_FUNCTION_TYPE,
+    KEY_ANNOTATIONS => TypeFactory.optional(TYPE_ANNOTATIONS),
+    KEY_FINAL => TypeFactory.optional(PBooleanType::DEFAULT),
+    KEY_OVERRIDE => TypeFactory.optional(PBooleanType::DEFAULT)
+  })
+  TYPE_FUNCTIONS = TypeFactory.hash_kv(TYPE_MEMBER_NAME, TypeFactory.not_undef)
+
+  TYPE_EQUALITY = TypeFactory.variant(TYPE_MEMBER_NAME, TypeFactory.array_of(TYPE_MEMBER_NAME))
+
+  TYPE_CHECKS = PAnyType::DEFAULT # TBD
+
+  TYPE_OBJECT_I12N = TypeFactory.struct({
+    KEY_PARENT => TypeFactory.optional(PType::DEFAULT),
+    KEY_ATTRIBUTES => TypeFactory.optional(TYPE_ATTRIBUTES),
+    KEY_FUNCTIONS => TypeFactory.optional(TYPE_FUNCTIONS),
+    KEY_EQUALITY => TypeFactory.optional(TYPE_EQUALITY),
+    KEY_EQUALITY_INCLUDE_TYPE => TypeFactory.optional(PBooleanType::DEFAULT),
+    KEY_CHECKS =>  TypeFactory.optional(TYPE_CHECKS),
+    KEY_ANNOTATIONS =>  TypeFactory.optional(TYPE_ANNOTATIONS)
+  })
+
+  # @abstract Encapsulates behavior common to {PAttribute} and {PFunction}
+  # @api public
+  class PAnnotatedMember
+
+    # @return [String] the name of this member
+    # @api public
+    attr_reader :name
+
+    # @return [PAnyType] the type of this member
+    # @api public
+    attr_reader :type
+
+    # @return [Hash{PType => Hash}] the annotations or `nil`
+    # @api public
+    attr_reader :annotations
+
+    # @param name [String] The name of the member
+    # @param [Hash{String=>Object}] i12n_hash
```

Comment:
the name of the param is the wrong place (should be before the type), and does not have a description